<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="VarSim">
    <meta name="author" content="Marghoob Mohiyuddin" >

    <title>SomaticSeq</title>

    <!-- Bootstrap core CSS -->
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .container {
        margin-right: auto;
        margin-left: auto;
        max-width: 760px;
        }
    </style>

    <script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
 (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
 m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
 })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-58409498-5', 'auto');
ga('send', 'pageview');


  /**
* Function that tracks a click on an outbound link in Google Analytics.
* This function takes a valid URL string as an argument, and uses that URL string
* as the event label.
*/
var trackOutboundLink = function(url) {
   ga('send', 'event', 'outbound', 'click', url, {'hitCallback':
     function () {
     document.location = url;
     }
   });
}


function _gaLt(event){
    var el = event.srcElement || event.target;

    /* Loop up the DOM tree through parent elements if clicked element is not a link (eg: an image inside a link) */
    while(el && (typeof el.tagName == 'undefined' || el.tagName.toLowerCase() != 'a' || !el.href)){
        el = el.parentNode;
    }

    if(el && el.href){
        if(el.href.indexOf(location.host) == -1){ /* external link */
            /* HitCallback function to either open link in either same or new window */
            var hitBack = function(link, target){
                target ? window.open(link, target) : window.location.href = link;
            };
            /* link */
            var link = el.href;
            /* Is target set and not _(self|parent|top)? */
            var target = (el.target && !el.target.match(/^_(self|parent|top)$/i)) ? el.target : false;
            /* send event with callback */
            ga(
                "send", "event", "Outgoing Links", link,
                document.location.pathname + document.location.search,
                {"hitCallback": hitBack(link, target)}
            );

            /* Prevent standard click */
            event.preventDefault ? event.preventDefault() : event.returnValue = !1;
        }

    }
}

/* Attach the event to all clicks in the document after page has loaded */
var w = window;
w.addEventListener ? w.addEventListener("load",function(){document.body.addEventListener("click",_gaLt,!1)},!1)
 : w.attachEvent && w.attachEvent("onload",function(){document.body.attachEvent("onclick",_gaLt)});


    </script>

</head>

<body>


<div class="container" style="padding-bottom: 400px;">
    <div class="page-header">
        <h1>SomaticSeq</h1>

        <h2>An ensemble approach to accurately detect somatic mutations</h2>
    </div>

    <h2>Publication [Open access]</h2>

    <div class="panel panel-default" style="font-family:monospace;">
        <div class="panel-body">
            <p>The SomaticSeq paper is published in <a href="http://dx.doi.org/10.1186/s13059-015-0758-2" target=_blank>Genome Biology (2015) 16:197</a>.</p>
            <p>SomaticSeq placed <a href="http://blog.bina.com/news/bina-places-1st-and-2nd-in-dream-challenge-establishing-potential-future-standards-for-cancer-genome-analysis" target=_blank>#1 and #2 in INDEL and SNV</a> in the Stage 5 of the <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/70726" target=_blank>ICGC-TCGA DREAM Somatic Mutation Calling Challenge</a>.</p>
            <p>Documentation can be directly downloaded <a href="https://github.com/bioinform/somaticseq/blob/master/docs/Manual.pdf?raw=true" target=_blank>here</a>.</p>
            <p>Watch the SomaticSeq Webinar On Demand<a href="http://info.bina.com/somaticseq-webinar-recording" target=_blank>here</a>.</p>
            <p>Data used in the manuscript is described <a href="data.html" target=_blank>here</a>.</p>
        </div>
    </div>




    <h2>Download SomaticSeq</h2>
    <ul>
    <li>The source code is hosted at <a href="https://github.com/bioinform/somaticseq/" onclick="trackOutboundLink('https://github.com/bioinform/somaticseq'); return false'">https://github.com/bioinform/somaticseq</a>
    </li>

    <li>The packaged releases can be found at <a href="https://github.com/bioinform/somaticseq/releases"
                                           onclick="trackOutboundLink('https://github.com/bioinform/somaticseq/releases'); return false;">https://github.com/bioinform/somaticseq/releases</a>
    </li>
    </ul>




    <h2>Software Dependencies</h2>
    <ul>
    <li>Python3 and regex package</li>
    <li>R and ada package</li>
    <li>SAMtools version 0.1.x (support for later versions is planned).</li>
    <li>GATK</li>
    <li>snpEFF and snpSIFT</li>
    <li>dbSNP and COSMIC file</li>
    </ul>




    <h2>Running SomaticSeq</h2>
    This SomaticSeq workflow takes about 3 hours for an ensmeble call set of 50K calls.

    The shell command (VCF file can also be bgzipped, but make sure it has .gz extention):
    
    <ul>
    <li><b>Running SomaticSeq's somatic mutation prediction mode:</b>
    <code>SomaticSeq.Wrapper.sh -M $PATH/TO/MuTect/variants.snv.vcf -V $PATH/TO/VarScan/variants.snv.vcf -J $PATH/TO/JointSNVMix/variants.snv.vcf -S $PATH/TO/SomaticSniper/variants.snv.vcf -D $PATH/TO/Vardict/variants.vcf -U $PATH/TO/MuSE/variants.snv.vcf -I $PATH/TO/Indelocator/variants.indel.vcf -v $PATH/TO/Varscan/variants.indel.vcf -N normal.bam -T tumor.bam -R ada_model_predictor.R -C sSNV.Classifier.RData -x sINDEL.Classifier.RData -g human_b37.fasta -c cosmic.b37.v71.vcf -d dbSNP.b37.v141.vcf -s $PATH/TO/DIR/snpSift -G $PATH/TO/GenomeAnalysisTK.jar -o $OUTPUT_DIR
    </code>
    </li>
    </ul>


    <ul>
    <li><b>Running SomaticSeq's model training mode:</b>
    <code>SomaticSeq.Wrapper.sh -M $PATH/TO/MuTect/variants.snv.vcf -V $PATH/TO/VarScan/variants.snv.vcf -J $PATH/TO/JointSNVMix/variants.snv.vcf -S $PATH/TO/SomaticSniper/variants.snv.vcf -D $PATH/TO/Vardict/variants.vcf -U $PATH/TO/MuSE/variants.snv.vcf -I $PATH/TO/Indelocator/variants.indel.vcf -v $PATH/TO/Varscan/variants.indel.vcf -N normal.bam -T tumor.bam -R ada_model_builder.R -g human_b37.fasta -c cosmic.b37.v71.vcf -d dbSNP.b37.v141.vcf -s $PATH/TO/DIR/snpSift -G $PATH/TO/GenomeAnalysisTK.jar -i Regions_to_Ignore.bed -Z SNV_Ground_Truth.vcf -z Indel_Ground_Truth.vcf -o $OUTPUT_DIR
    </code>
    </li>
    </ul>



    <h3>Command-line options for SomaticSeq.Wrapper.sh:</h3>
    <ul>
    <li><code>-M variants.vcf</code>: MuTect's output VCF (Optional)</li>
    <li><code>-V variants.snp.vcf</code>: VarScan2's output SNV VCF (Optional)</li>
    <li><code>-J variants.vcf</code>: JointSNVMix's variant output converted to VCF (Optional)</li>
    <li><code>-S variants.vcf</code>: SomaticSniper's output VCF (Optional)</li>
    <li><code>-D variants.vcf</code>: VarDict's output VCF (Optional)</li>
    <li><code>-U variants.vcf</code>: MuSE's output VCF (Optional)</li>
    <li><code>-I variants.vcf</code>: Indelocator's output VCF (Optional)</li>
    <li><code>-v variants.indel.vcf</code>: VarScan2's output INDEL VCF (Optional)</li>
    <li><code>-N normal.bam</code>: Normal BAM files</li>
    <li><code>-T tumor.bam</code>: Tumor BAM files</li>
    <li><code>-R ada_model_[predictor|builder].R</code>: Ada script in R. (Optional. If not provided, SomaticSeq will not attempt to train or predict. If -C and -x are provided, this script is assumed to be predictor.R. If -Z and -z are provided, this script is assumed to be builder.R).</li>
    <li><code>-C Trained_SNV_Classifier.RData</code>: Trained model/classifier for SNV, e.g., from DREAM Challenge (Optional: if provided along with predictor.R, SomaticSeq will predict somatic mutations.)</li>
    <li><code>-x Trained_INDEL_Classifier.RData</code>: Trained model/classifier for INDEL, e.g., from DREAM Challenge (Optional: if provided along with predictor.R, SomaticSeq will predict somatic mutations.)</li>
    <li><code>-g human_b37_decoy.fasta</code>: Genome reference FASTA file</li>
    <li><code>-c COSMIC.b37.vcf</code>: COSMIC VCF file</li>
    <li><code>-d dbSNP.v141.b37.vcf</code>: dbSNP VCF file</li>
    <li><code>-s path/to/SnpEFF</code>: snpEFF/snpSift's installation directory containing the executable .jar files</li>
    <li><code>-G path/to/GATK.jar</code>: GATK's java executable file</li>
    <li><code>-i IGNORE.bed</code>: Regions to toss out in the final VCF files (Optional)</li>
    <li><code>-Z SNV_Ground_Truth.vcf</code>: A VCF file containing true sSNVs, and everything else is true negative. (Optional. If provided with builder.R, SomaticSeq will train the model.)</li>
    <li><code>-z INDEL_Ground_Truth.vcf</code>: A VCF file containing true sINDELs, and everything else is true negative. (Optional. If provided with builder.R, SomaticSeq will train the model.)</li>
    <li><code>-o output</code>: Output directory (make sure it exists)</li>
    </ul>



    <h3>Known Issues:</h3>
    Some older versions of GATK seem to have problem with FIFO used in the SomaticSeq script (i.e., we had problem with GATK version 2014.1-2.8.1), so we recommend at least version version 2014.4-2 or later.



    <h3>Contact Us:</h3>
    Please report bugs and suggestions to <a href="mailto:li_tai.fang@bina.roche.com">li_tai.fang@bina.roche.com</a>.




    <h3>References/Tools</h3>
    <ul>
    <li>MuTect: Cibulskis K, Lawrence MS, Carter SL, et al. Sensitive detection of somatic point mutations in impure and heterogeneous cancer samples. Nat Biotechnol. 2013;31(3):213-9.</li>
    <li>VarScan2: Koboldt DC, Zhang Q, Larson DE, et al. VarScan 2: somatic mutation and copy number alteration discovery in cancer by exome sequencing. Genome Res. 2012;22(3):568-76.</li>
    <li>JointSNVMix2: Roth A, Ding J, Morin R, et al. JointSNVMix: a probabilistic model for accurate detection of somatic mutations in normal/tumour paired next-generation sequencing data. Bioinformatics. 2012;28(7):907-13.</li>
    <li>SomaticSniper: Larson DE, Harris CC, Chen K, et al. SomaticSniper: identification of somatic point mutations in whole genome sequencing data. Bioinformatics. 2012;28(3):311-7.</li>
    <li>VarDict: Lai, Z., Markovets, A., Ahdesmaki, M., Johnson, J.: VarDict: A Novel and Versatile Variant Caller for Next-generation Sequencing in Cancer Research. 2015 AACR Abstract</li>
    <li>Indelocator: Banerji S, Cibulskis K, Rangel-escareno C, et al. Sequence analysis of mutations and translocations across breast cancer subtypes. Nature. 2012;486(7403):405-9.</li>
    <li>MuSE: <a href="http://bioinformatics.mdanderson.org/main/MuSE" target=_blank>http://bioinformatics.mdanderson.org/main/MuSE</a></li>
    </ul>




</div>


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
</body>
