<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="VarSim">
    <meta name="author" content="Marghoob Mohiyuddin" >

    <title>SomaticSeq</title>

    <!-- Bootstrap core CSS -->
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .container {
        margin-right: auto;
        margin-left: auto;
        max-width: 760px;
        }
    </style>

    <script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
 (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
 m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
 })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-58409498-5', 'auto');
ga('send', 'pageview');


  /**
* Function that tracks a click on an outbound link in Google Analytics.
* This function takes a valid URL string as an argument, and uses that URL string
* as the event label.
*/
var trackOutboundLink = function(url) {
   ga('send', 'event', 'outbound', 'click', url, {'hitCallback':
     function () {
     document.location = url;
     }
   });
}


function _gaLt(event){
    var el = event.srcElement || event.target;

    /* Loop up the DOM tree through parent elements if clicked element is not a link (eg: an image inside a link) */
    while(el && (typeof el.tagName == 'undefined' || el.tagName.toLowerCase() != 'a' || !el.href)){
        el = el.parentNode;
    }

    if(el && el.href){
        if(el.href.indexOf(location.host) == -1){ /* external link */
            /* HitCallback function to either open link in either same or new window */
            var hitBack = function(link, target){
                target ? window.open(link, target) : window.location.href = link;
            };
            /* link */
            var link = el.href;
            /* Is target set and not _(self|parent|top)? */
            var target = (el.target && !el.target.match(/^_(self|parent|top)$/i)) ? el.target : false;
            /* send event with callback */
            ga(
                "send", "event", "Outgoing Links", link,
                document.location.pathname + document.location.search,
                {"hitCallback": hitBack(link, target)}
            );

            /* Prevent standard click */
            event.preventDefault ? event.preventDefault() : event.returnValue = !1;
        }

    }
}

/* Attach the event to all clicks in the document after page has loaded */
var w = window;
w.addEventListener ? w.addEventListener("load",function(){document.body.addEventListener("click",_gaLt,!1)},!1)
 : w.attachEvent && w.attachEvent("onload",function(){document.body.attachEvent("onclick",_gaLt)});


    </script>

</head>

<body>


<div class="container" style="padding-bottom: 400px;">
    <div class="page-header">
        <h1>SomaticSeq</h1>

        <h2>An ensemble approach to accurately detect somatic mutations</h2>
    </div>

    <h2>Publication [Open access]</h2>

    <div class="panel panel-default" style="font-family:monospace;">
        <div class="panel-body">

            <p>If you use SomaticSeq in your work, please cite the following:<br>
                <ul>
                <li>
                <small>
                Li Tai Fang, Pegah Tootoonchi Afshar, Aparna Chhibber, Marghoob Mohiyuddin, Yu Fan, John C. Mu, Greg Gibeling, Sharon Barr, Narges Bani Asadi, Mark B. Gerstein, Daniel C. Koboldt, Wenyi Wang, Wing H. Wong, and Hugo Y.K. Lam.
                An Ensemble Approach to Accurately Detect Somatic Mutations Using SomaticSeq. 
                <i>Genome Biology, 16(1):197. (2015).</i>
                <a href="http://dx.doi.org/10.1186/s13059-015-0758-2" target=_blank>DOI: 10.1186/s13059-015-0758-2</a>
                </small>
                </li>
                </ul>
            </p>

            <p>SomaticSeq placed <a href="http://sequencing.roche.com/news---media/press-releases/bina-places-1st-and-2nd-in-dream-challenge.html" target=_blank>#1 and #2 in INDEL and SNV</a> in the Stage 5 of the <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/70726" target=_blank>ICGC-TCGA DREAM Somatic Mutation Calling Challenge</a>.</p>

            <p>Watch the quick <b><a href="https://www.youtube.com/watch?v=MnJdTQWWN6w" target=_blank>8-minute youtube video</a></b> describing SomaticSeq,
               or the <a href="https://www.youtube.com/watch?v=jdwYkjE0Fzg" target=_blank>SomaticSeq Webinar On Demand</a>.</p>

            <p>Documentation can be directly downloaded <a href="https://github.com/bioinform/somaticseq/blob/master/docs/Manual.pdf?raw=true" target=_blank>here</a>.</p>

            <p>Data used in the manuscript is described <a href="data.html" target=_blank>here</a>.</p>

        </div>
    </div>




    <h2>Download SomaticSeq</h2>
    <ul>
    <li>The source code is hosted at <a href="https://github.com/bioinform/somaticseq/" onclick="trackOutboundLink('https://github.com/bioinform/somaticseq'); return false'">https://github.com/bioinform/somaticseq</a>
    </li>

    <li>The packaged releases can be found at <a href="https://github.com/bioinform/somaticseq/releases"
                                           onclick="trackOutboundLink('https://github.com/bioinform/somaticseq/releases'); return false;">https://github.com/bioinform/somaticseq/releases</a>
    </li>
    </ul>




    <h2>Software Dependencies</h2>
    <ul>
    <li>Python3, plus regex, numpy, scipy, and pysam packages</li>
    <li>R and ada package</li>
    <li>BEDtools if inclusion and/or exclusion regions are used</li>
    <li>Optional: free GATK (We use GATK CombineVariants. You may find other ways to combine VCF files.)</li>
    <li>Optional: dbSNP and COSMIC file (if you want to use those information as a part of your training set)</li>
    </ul>




    <h2>Running SomaticSeq</h2>
    This SomaticSeq workflow takes about 3 hours for an ensmeble call set of 50K calls.

    The shell command (All VCF files need to be properly sorted. VCF files can also be bgzipped, but make sure they have .gz extension).<br>
    
    Long options shown here are only available for v2.0.3 or later.
    
    <ul>
    <li><b>Running SomaticSeq's somatic mutation prediction mode:</b><br>
    
    <code>
    SomaticSeq.Wrapper.sh                                 \<br>
    --mutect $PATH/TO/MuTect/variants.snv.vcf             \<br>
    --mutect2 $PATH/TO/MuTect2/variants.vcf               \<br>
    --indelocator $PATH/TO/Indelocator/variants.indel.vcf \<br>
    --varscan-snv $PATH/TO/VarScan/variants.snv.vcf       \<br>
    --varscan-indel $PATH/TO/Varscan/variants.indel.vcf   \<br>
    --jsm $PATH/TO/JointSNVMix/variants.snv.vcf           \<br>
    --sniper $PATH/TO/SomaticSniper/variants.snv.vcf      \<br>
    --vardict $PATH/TO/Vardict/variants.vcf               \<br>
    --muse $PATH/TO/MuSE/variants.snv.vcf                 \<br>
    --lofreq-snv    $PATH/TO/LoFreq/variants.snp.vcf      \<br>
    --lofreq-indel  $PATH/TO/LoFreq/variants.indel.vcf    \<br>
    --scalpel $PATH/TO/Scalpel/variants.indel.vcf         \<br>
    --strelka-snv $PATH/TO/Strelka/variants.snv.vcf       \<br>
    --strelka-indel $PATH/TO/Strelka/variants.indel.vcf   \<br>
    --normal-bam normal.bam                               \<br>
    --tumor-bam tumor.bam                                 \<br>
    --ada-r-script ada_model_predictor.R                  \<br>
    --classifier-snv sSNV.Classifier.RData                \<br>
    --classifier-indel sINDEL.Classifier.RData            \<br>
    --pass-threshold 0.5                                  \<br>
    --lowqual-threshold 0.1                               \<br>
    --genome-reference human_b37.fasta                    \<br>
    --cosmic cosmic.b37.v71.vcf                           \<br>
    --dbsnp dbSNP.b37.v141.vcf                            \<br>
    --inclusion-region validated.bed                      \<br>
    --exclusion-region blacklist.bed                      \<br>
    --gatk $PATH/TO/GenomeAnalysisTK.jar                  \<br>
    --output-dir $OUTPUT_DIR
    </code>
    
    </li>
    </ul>


    <ul>
    <li><b>Running SomaticSeq's model training mode:</b><br>
    
    <code>
    SomaticSeq.Wrapper.sh                                 \<br>
    --mutect $PATH/TO/MuTect/variants.snv.vcf             \<br>
    --mutect2 $PATH/TO/MuTect2/variants.vcf               \<br>
    --indelocator $PATH/TO/Indelocator/variants.indel.vcf \<br>
    --varscan-snv $PATH/TO/VarScan/variants.snv.vcf       \<br>
    --varscan-indel $PATH/TO/Varscan/variants.indel.vcf   \<br>
    --jsm $PATH/TO/JointSNVMix/variants.snv.vcf           \<br>
    --sniper $PATH/TO/SomaticSniper/variants.snv.vcf      \<br>
    --vardict $PATH/TO/Vardict/variants.vcf               \<br>
    --muse $PATH/TO/MuSE/variants.snv.vcf                 \<br>
    --lofreq-snv    $PATH/TO/LoFreq/variants.snp.vcf      \<br>
    --lofreq-indel  $PATH/TO/LoFreq/variants.indel.vcf    \<br>
    --scalpel $PATH/TO/Scalpel/variants.indel.vcf         \<br>
    --strelka-snv $PATH/TO/Strelka/variants.snv.vcf       \<br>
    --strelka-indel $PATH/TO/Strelka/variants.indel.vcf   \<br>
    --normal-bam normal.bam                               \<br>
    --tumor-bam tumor.bam                                 \<br>
    --ada-r-script ada_model_builder.R                    \<br>
    --genome-reference human_b37.fasta                    \<br>
    --cosmic cosmic.b37.v71.vcf                           \<br>
    --dbsnp dbSNP.b37.v141.vcf                            \<br>
    --gatk $PATH/TO/GenomeAnalysisTK.jar                  \<br>
    --inclusion-region validated.bed                      \<br>
    --exclusion-region blacklist.bed                      \<br>
    --truth-snv SNV_Ground_Truth.vcf                      \<br>
    --truth-indel Indel_Ground_Truth.vcf                  \<br>
    --output-dir $OUTPUT_DIR
    </code>
    </li>
    </ul>



    <h3>Command-line options for SomaticSeq.Wrapper.sh:</h3>
    <h4>Long options only available after v2.0.3</h4>
    <ul>
    <li><code>-M | --mutect</code>: MuTect's output VCF (Optional)</li>
    <li><code>-I | --indelocator</code>: Indelocator's output VCF (Optional)</li>
    <li><code>-2 | --mutect2</code>: MuTect2's output VCF (Optional)</li>
    <li><code>-V | --varscan-snv</code>: VarScan2's output SNV VCF (Optional)</li>
    <li><code>-v | --varscan-indel</code>: VarScan2's output INDEL VCF (Optional)</li>
    <li><code>-J | --jsm</code>: JointSNVMix's variant output converted to VCF (Optional)</li>
    <li><code>-S | --sniper</code>: SomaticSniper's output VCF (Optional)</li>
    <li><code>-D | --vardict</code>: VarDict's output VCF (Optional)</li>
    <li><code>-U | --muse</code>: MuSE's output VCF (Optional)</li>
    <li><code>-L | --lofreq-snv</code>: LoFreq's output VCF (Optional)</li>
    <li><code>-l | --lofreq-indel</code>: LoFreq's output INDEL VCF (Optional)</li>
    <li><code>-p | --scalpel</code>: Scalpel's output VCF (Optional)</li>
    <li><code>--strelka-snv</code>: Strelka2's SNV VCF (Optional)</li>
    <li><code>--strelka-indel</code>: Strelka2's INDEL VCF (Optional)</li>
    <li><code>-N | --normal-bam</code>: Normal BAM files</li>
    <li><code>-T | --tumor-bam</code>: Tumor BAM files</li>
    <li><code>-R | --ada-r-script</code>: Ada script in R. (Optional. If not provided, SomaticSeq will not attempt to train or predict. If -C and -x are provided, this script is assumed to be predictor.R. If -Z and -z are provided, this script is assumed to be builder.R).</li>
    <li><code>-C | --classifier-snv</code>: Trained model/classifier for SNV, e.g., from DREAM Challenge (Optional: if provided along with predictor.R, SomaticSeq will predict somatic mutations.)</li>
    <li><code>-x | --classifier-indel</code>: Trained model/classifier for INDEL, e.g., from DREAM Challenge (Optional: if provided along with predictor.R, SomaticSeq will predict somatic mutations.)</li>
    <li><code>-g | --genome-reference</code>: Genome reference FASTA file</li>
    <li><code>-c | --cosmic</code>: COSMIC VCF file</li>
    <li><code>-d | --dbsnp</code>: dbSNP VCF file</li>
    <li><code>-G | --gatk</code>: GATK's java executable file</li>
    <li><code>-i | --inclusion-region</code>: Only include those regions in analysis</li>
    <li><code>-e | --exclusion-region</code>: Regions to toss out in the final VCF files (Optional)</li>
    <li><code>-Z | --truth-snv</code>: A VCF file containing true sSNVs, and everything else is true negative. (Optional. If provided with builder.R, SomaticSeq will train the model.)</li>
    <li><code>-z | --truth-indel</code>: A VCF file containing true sINDELs, and everything else is true negative. (Optional. If provided with builder.R, SomaticSeq will train the model.)</li>
    <li><code>-o | --output-dir</code>: Output directory (make sure it exists)</li>
    <li><code>-k | --keep-intermediates</code>: [0|1], whether to keep intermediate files for debugging. [Optional. Default = 0]</li>
    <li><code>--pass-threshold</code>: Probability value above which a call will be labeled as PASS. (Default = 0.5. Optional. Available after v2.2.2)</li>
    <li><code>--lowqual-threshold</code>: Probability value above which (but below the PASS threshold) a call will be labeled as LowQual. (Default = 0.1. Optional. Available after v2.2.2)</li>
    </ul>



    <h3>Contact Us</h3>
    Please report bugs and suggestions to the <a href="https://github.com/bioinform/somaticseq/issues" target=_blank>report issues</a> page. The developers will be notified immediately. 
    Alternatively, you may email <a href="mailto:li_tai.fang@bina.roche.com">li_tai.fang@bina.roche.com</a>.



    <h3>References/Tools</h3>
    <ul>

    <li>MuTect:        <a href="http://dx.doi.org/10.1038/nbt.2514" target=_blank>Cibulskis K, Lawrence MS, Carter SL, et al. Sensitive detection of somatic point mutations in impure and heterogeneous cancer samples. Nat Biotechnol. 2013;31(3):213-9.</a></li>
    <li>VarScan2:      <a href="http://dx.doi.org/10.1101/gr.129684.111" target=_blank>Koboldt DC, Zhang Q, Larson DE, et al. VarScan 2: somatic mutation and copy number alteration discovery in cancer by exome sequencing. Genome Res. 2012;22(3):568-76.</a></li>
    <li>JointSNVMix2:  <a href="http://dx.doi.org/10.1093/bioinformatics/bts053" target=_blank>Roth A, Ding J, Morin R, et al. JointSNVMix: a probabilistic model for accurate detection of somatic mutations in normal/tumour paired next-generation sequencing data. Bioinformatics. 2012;28(7):907-13.</a></li>
    <li>SomaticSniper: <a href="http://dx.doi.org/10.1093/bioinformatics/btr665" target=_blank> Larson DE, Harris CC, Chen K, et al. SomaticSniper: identification of somatic point mutations in whole genome sequencing data. Bioinformatics. 2012;28(3):311-7.</a></li>
    <li>VarDict:       <a href="http://dx.doi.org/10.1093/nar/gkw227" target=_blank>Lai Z, Markovets A, Ahdesmaki M, Chapman B, et al. VarDict: a novel and versatile variant caller for next-generation sequencing in cancer research. Nucleic Acids Research. 2016.</a></li>
    <li>Indelocator:   <a href="http://dx.doi.org/10.1038/nature11154" target=_blank'>Banerji S, Cibulskis K, Rangel-escareno C, et al. Sequence analysis of mutations and translocations across breast cancer subtypes. Nature. 2012;486(7403):405-9.</a></li>
    <li>MuSE:          <a href="http://dx.doi.org/10.1186/s13059-016-1029-6" target=_blank>Fan Y, Xi L, Hughes DS, et al. MuSE: accounting for tumor heterogeneity using a sample-specific error model improves sensitivity and specificity in mutation calling from sequencing data. Genome Biol. 2016;17(1):178.</a></li>
    <li>LoFreq:        <a href="http://dx.doi.org/10.1093/nar/gks918" target=_blank>Wilm A, Aw PP, Bertrand D, et al. LoFreq: a sequence-quality aware, ultra-sensitive variant caller for uncovering cell-population heterogeneity from high-throughput sequencing datasets. Nucleic Acids Res. 2012;40(22):11189-201.</a>
    <li>Scalpel:       <a href="http://dx.doi.org/10.1038/nmeth.3069" target=_blank>Narzisi G, O'rawe JA, Iossifov I, et al. Accurate de novo and transmitted indel detection in exome-capture data using microassembly. Nat Methods. 2014;11(10):1033-6.</a>
    <li>Strelka2:      <a href="http://dx.doi.org/10.1093/bioinformatics/bts271" target=_blank>Saunders CT, Wong WS, Swamy S, Becq J, Murray LJ, Cheetham RK. Strelka: accurate somatic small-variant calling from sequenced tumor-normal sample pairs. Bioinformatics. 2012;28(14):1811-7.</a>
    </ul>


    <div style="text-align: center;">

    <p>
    <br><br>
    <h3>A short video describing SomaticSeq:</h3>
    
    <iframe width="640" height="360" src="https://www.youtube.com/embed/MnJdTQWWN6w" frameborder="0" allowfullscreen></iframe>
    </p>
    </div>


</div>


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
</body>
