<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="VarSim">
    <meta name="author" content="Marghoob Mohiyuddin" >

    <title>SomaticSeq</title>

    <!-- Bootstrap core CSS -->
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .container {
        margin-right: auto;
        margin-left: auto;
        max-width:960px;
        }
    </style>

    <script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
 (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
 m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
 })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-58409498-5', 'auto');
ga('send', 'pageview');


  /**
* Function that tracks a click on an outbound link in Google Analytics.
* This function takes a valid URL string as an argument, and uses that URL string
* as the event label.
*/
var trackOutboundLink = function(url) {
   ga('send', 'event', 'outbound', 'click', url, {'hitCallback':
     function () {
     document.location = url;
     }
   });
}


function _gaLt(event){
    var el = event.srcElement || event.target;

    /* Loop up the DOM tree through parent elements if clicked element is not a link (eg: an image inside a link) */
    while(el && (typeof el.tagName == 'undefined' || el.tagName.toLowerCase() != 'a' || !el.href)){
        el = el.parentNode;
    }

    if(el && el.href){
        if(el.href.indexOf(location.host) == -1){ /* external link */
            /* HitCallback function to either open link in either same or new window */
            var hitBack = function(link, target){
                target ? window.open(link, target) : window.location.href = link;
            };
            /* link */
            var link = el.href;
            /* Is target set and not _(self|parent|top)? */
            var target = (el.target && !el.target.match(/^_(self|parent|top)$/i)) ? el.target : false;
            /* send event with callback */
            ga(
                "send", "event", "Outgoing Links", link,
                document.location.pathname + document.location.search,
                {"hitCallback": hitBack(link, target)}
            );

            /* Prevent standard click */
            event.preventDefault ? event.preventDefault() : event.returnValue = !1;
        }

    }
}

/* Attach the event to all clicks in the document after page has loaded */
var w = window;
w.addEventListener ? w.addEventListener("load",function(){document.body.addEventListener("click",_gaLt,!1)},!1)
 : w.attachEvent && w.attachEvent("onload",function(){document.body.attachEvent("onclick",_gaLt)});


    </script>

</head>

<body style="color: #D8D8D8; background-color: #000000;" link="#00FFFF" vlink="#FB00FF" alink="#EE0000">


<div class="container" style="padding-bottom: 400px;">
    <div class="page-header">
        <h1>SomaticSeq</h1>

        <h2>An ensemble approach to accurately detect somatic mutations</h2>
    </div>

    <h2>Publication [Open access]</h2>

    <div class="panel panel-default" style="color: #D8D8D8; background-color: #000000;; font-family:monospace;">
        <div class="panel-body">

            <p>If you use SomaticSeq in your work, please cite the following:<br>
                <ul>
                <li>
                    Fang LT, Afshar PT, Chhibber A, Mohiyuddin M, Fan Y, Mu JC, Gibeling G, Barr S, Bani Asadi N, Gerstein MB, <i>et al</i>.
                    An ensemble approach to accurately detect somatic mutations using SomaticSeq.
                    <i>Genome Biology. 2015;16(1):197. <a href="http://dx.doi.org/10.1186/s13059-015-0758-2" target="_blank">doi:10.1186/s13059-015-0758-2</a></i>
                </li>
                </ul>
            </p>

            <p>
            The <a href="https://www.fda.gov/science-research/bioinformatics-tools/microarraysequencing-quality-control-maqcseqc">SEQC2/MAQC-IV Consortium</a> has published <a href="https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP162370">numerous whole-genome and whole-exome sequencing replicates from multiple sequencing centers</a> for a pair of tumor-normal reference samples, along with the <a href="https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release/latest/">high-confidence somatic mutation reference call set</a>. These resources can be used to train machine learning classifers (e.g., <a href="https://doi.org/10.1101/667261">Sahraeian S.M.E. <i>et al</i>. bioRxiv 2019</a>) or evaluate algorithms and pipelines (e.g., <a href="https://doi.org/10.1038/s41587-021-00994-5">Xiao W. <i>et al</i>. Nat Biotechnol 2021</a>). That work is published as:<br>

                <ul>
                <li>
                    Fang LT, Zhu B, Zhao Y, Chen W, Yang Z, Kerrigan L, Langenbach K, de Mars M, Lu C, Idler K, <i>et al</i>.
                    Establishing community reference samples, data and call sets for benchmarking cancer mutation detection using whole-genome sequencing.
                    <i>Nature Biotechnology. 2021;39(9):1151-1160. <a href="http://doi.org/10.1038/s41587-021-00993-6" target="_blank">doi:10.1038/s41587-021-00993-6</a></i>. 
                </li>
                </ul>

            </p>

            <p>SomaticSeq was the tool that had Bina Technologies, Inc. ranked <a href="http://sequencing.roche.com/content/rochesequence/en/news-overview/2015/bina-places-1st-and-2nd-in-dream-challenge.html" target=_blank>#1 and #2 in INDEL and SNV</a> in the Stage 5 of the <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/70726" target=_blank>ICGC-TCGA DREAM Somatic Mutation Calling Challenge</a>.</p>

            <p>Watch the quick <b><a href="https://www.youtube.com/watch?v=MnJdTQWWN6w" target=_blank>8-minute youtube video</a></b> describing SomaticSeq.</p>

            <p>Documentation can be directly downloaded <a href="https://github.com/bioinform/somaticseq/blob/master/docs/Manual.pdf?raw=true" target=_blank>here</a>.</p>

        </div>
    </div>




    <h2>Download SomaticSeq</h2>

    <ul>
        <li>The source code is hosted at <a href="https://github.com/bioinform/somaticseq/" onclick="trackOutboundLink('https://github.com/bioinform/somaticseq'); return false'">https://github.com/bioinform/somaticseq</a></li>

        <li>The packaged releases can be found at <a href="https://github.com/bioinform/somaticseq/releases"
                                           onclick="trackOutboundLink('https://github.com/bioinform/somaticseq/releases'); return false;">https://github.com/bioinform/somaticseq/releases</a></li>

        <li>Docker image can be downloaded at <a href="https://hub.docker.com/r/lethalfang/somaticseq/" onclick="trackOutboundLink('https://hub.docker.com/r/lethalfang/somaticseq/'); return false'">https://hub.docker.com/r/lethalfang/somaticseq/</a>
    </ul>




    <h2>Software Dependencies</h2>

    <ul>
        <li>Python3, plus numpy, scipy, and pysam packages</li>
        <li>R and ada package</li>
        <li>BEDtools if inclusion and/or exclusion regions are used</li>
        <li>Optional: dbSNP and COSMIC VCF file (if you want to use those information as a part of your training set)</li>
    </ul>



    <h2>Running SomaticSeq</h2>

        <ul>
            <li>This SomaticSeq procedure takes about 3 hours for a call set of 50K calls on 30X tumor-normal genome on a single thread.</li>
            <li>All VCF, BAM, and FASTA files need to be sorted identically. For VCF files, both .vcf and .vcf.gz are acceptable.</li>
            <li>The following command can be run when those individual mutation callers have completed running. 
                To use our entire somatic mutation pipelines, read the README in our <a href="https://github.com/bioinform/somaticseq/" onclick="trackOutboundLink('https://github.com/bioinform/somaticseq'); return false'">source code</a>. The pipeline is available in both <a href="https://www.docker.com/" target=_blank>Docker</a> and <a href="http://singularity.lbl.gov/" target=_blank>Singularity</a>.</li>
        </ul>
    



            <h3>Example command to merge caller results and extract SomaticSeq features</h3>
            
            
            <h4>Tumor-normal <b>paired</b> mode</h4> 
            
                <ul>
                    <code>
                        somaticseq_parallel.py
                        --output-directory  $OUTPUT_DIR
                        --genome-reference  GRCh38.fa
                        --inclusion-region  genome.bed
                        --exclusion-region  blacklist.bed
                        <b>paired</b>
                        --tumor-bam-file    tumor.bam
                        --normal-bam-file   matched_normal.bam
                        --mutect2-vcf       MuTect2/variants.vcf
                        --varscan-snv       VarScan2/variants.snp.vcf
                        --varscan-indel     VarScan2/variants.indel.vcf
                        --jsm-vcf           JointSNVMix2/variants.snp.vcf
                        --somaticsniper-vcf SomaticSniper/variants.snp.vcf
                        --vardict-vcf       VarDict/variants.vcf
                        --muse-vcf          MuSE/variants.snp.vcf
                        --lofreq-snv        LoFreq/variants.snp.vcf
                        --lofreq-indel      LoFreq/variants.indel.vcf
                        --scalpel-vcf       Scalpel/variants.indel.vcf
                        --strelka-snv       Strelka/variants.snv.vcf
                        --strelka-indel     Strelka/variants.indel.vcf
                    </code>
                </ul>
            



            <h4>Tumor-only <b>single</b> mode</h4> 
            
                <ul>
                    <code>
                        somaticseq_parallel.py
                        --output-directory  $OUTPUT_DIR
                        --genome-reference  GRCh38.fa
                        --inclusion-region  genome.bed
                        --exclusion-region  blacklist.bed
                        <b>single</b>
                        --bam-file          tumor.bam
                        --mutect2-vcf       MuTect2/variants.vcf
                        --varscan-vcf       VarScan2/variants.vcf
                        --vardict-vcf       VarDict/variants.vcf
                        --lofreq-vcf        LoFreq/variants.vcf
                        --scalpel-vcf       Scalpel/variants.indel.vcf
                        --strelka-vcf       Strelka/variants.vcf
                    </code>
                </ul>
            



            <h4>Additional parameters to be specified before <b>paired</b> or <b>single</b> option to invoke training mode.</h4>
            
                <ul>
                    <li><code>--somaticseq-train</code>: FLAG to invoke training mode with no argument, which also requires the following inputs, R and ada package in R</li>
                    <li><code>--truth-snv</code>: if you have ground truth VCF file for SNV</li>
                    <li><code>--truth-indel</code>: if you have a ground truth VCF file for INDEL</li>
                </ul>



            <h4>Additional input files to be specified before <b>paired</b> or <b>single</b> option invoke prediction mode (to use classifiers to score variants)</h4>

                <ul>
                    <li><code>--classifier-snv</code>: classifier (.RData file) previously built for SNV</li>
                    <li><code>--classifier-indel</code>: classifier (.RData file) previously built for INDEL</li>
                </ul>



            
            <h4>Notes</h4> 

                <ul>
                    
                    <li>Without the paramters above to invoke training or prediction mode, SomaticSeq will default to majority-vote consensus mode.</li>
                    
                    <li><code>--inclusion-region</code> and/or <code>--exclusion-region</code> will require BEDTools in your path.</li>
    
                    <li>To split the job into multiple threads, place <code>--threads X</code> before the paired option to indicate X threads. It simply creates multiple BED file (each consisting of 1/X of total base pairs) for SomaticSeq to run on each of those sub-BED files in parallel. It then merges the results. This requires bedtools in your path.</li>
    
                    <li>For all input VCF files, either .vcf or .vcf.gz are acceptable.</li>
                </ul>

            




    <div style="text-align: center;">

    <p>
    <br><br>
    <h3>A short video describing SomaticSeq:</h3>
    
    <iframe width="640" height="360" src="https://www.youtube.com/embed/MnJdTQWWN6w" frameborder="0" allowfullscreen></iframe>
    </p>
    </div>



    <h3>Contact Us</h3>
    Please report bugs and suggestions to the <a href="https://github.com/bioinform/somaticseq/issues" target=_blank>report issues</a> page. The developers will be notified immediately. 
    Alternatively, you may email <a href="mailto:ltfang@gmail.com">ltfang@gmail.com</a>.


    <h3>References/Tools</h3>

    <ul>
        <li>MuTect:        <a href="http://dx.doi.org/10.1038/nbt.2514"              target=_blank>Cibulskis K, Lawrence MS, Carter SL, et al. Sensitive detection of somatic point mutations in impure and heterogeneous cancer samples. Nat Biotechnol. 2013;31(3):213-9.</a></li>
        <li>VarScan2:      <a href="http://dx.doi.org/10.1101/gr.129684.111"         target=_blank>Koboldt DC, Zhang Q, Larson DE, et al. VarScan 2: somatic mutation and copy number alteration discovery in cancer by exome sequencing. Genome Res. 2012;22(3):568-76.</a></li>
        <li>JointSNVMix2:  <a href="http://dx.doi.org/10.1093/bioinformatics/bts053" target=_blank>Roth A, Ding J, Morin R, et al. JointSNVMix: a probabilistic model for accurate detection of somatic mutations in normal/tumour paired next-generation sequencing data. Bioinformatics. 2012;28(7):907-13.</a></li>
        <li>SomaticSniper: <a href="http://dx.doi.org/10.1093/bioinformatics/btr665" target=_blank> Larson DE, Harris CC, Chen K, et al. SomaticSniper: identification of somatic point mutations in whole genome sequencing data. Bioinformatics. 2012;28(3):311-7.</a></li>
        <li>VarDict:       <a href="http://dx.doi.org/10.1093/nar/gkw227"            target=_blank>Lai Z, Markovets A, Ahdesmaki M, Chapman B, et al. VarDict: a novel and versatile variant caller for next-generation sequencing in cancer research. Nucleic Acids Research. 2016.</a></li>
        <li>Indelocator:   <a href="http://dx.doi.org/10.1038/nature11154"           target=_blank'>Banerji S, Cibulskis K, Rangel-escareno C, et al. Sequence analysis of mutations and translocations across breast cancer subtypes. Nature. 2012;486(7403):405-9.</a></li>
        <li>MuSE:          <a href="http://dx.doi.org/10.1186/s13059-016-1029-6"     target=_blank>Fan Y, Xi L, Hughes DS, et al. MuSE: accounting for tumor heterogeneity using a sample-specific error model improves sensitivity and specificity in mutation calling from sequencing data. Genome Biol. 2016;17(1):178.</a></li>
        <li>LoFreq:        <a href="http://dx.doi.org/10.1093/nar/gks918"            target=_blank>Wilm A, Aw PP, Bertrand D, et al. LoFreq: a sequence-quality aware, ultra-sensitive variant caller for uncovering cell-population heterogeneity from high-throughput sequencing datasets. Nucleic Acids Res. 2012;40(22):11189-201.</a>
        <li>Scalpel:       <a href="http://dx.doi.org/10.1038/nmeth.3069"            target=_blank>Narzisi G, O'rawe JA, Iossifov I, et al. Accurate de novo and transmitted indel detection in exome-capture data using microassembly. Nat Methods. 2014;11(10):1033-6.</a>
        <li>Strelka2:      <a href="https://doi.org/10.1038/s41592-018-0051-x" target=_blank>Kim S, Scheffler K, Halpern AL, et al. Strelka2: fast and accurate calling of germline and somatic variants. Nat Methods. 2018;15(8):591-594.</a>
    </ul>



</div>


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
</body>
