<!DOCTYPE html>
<html lang="en">

<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="VarSim">
    <meta name="author" content="Marghoob Mohiyuddin" >

    <title>SomaticSeq</title>

    <!-- Bootstrap core CSS -->
    <link href="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/css/bootstrap.min.css" rel="stylesheet">
    <style>
        .container {
        margin-right: auto;
        margin-left: auto;
        max-width: 760px;
        }
    </style>

    <script>
(function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
 (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
 m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
 })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

ga('create', 'UA-58409498-5', 'auto');
gq('require', 'linkid', 'linkid.js');
ga('send', 'pageview');
  
  
  /**
* Function that tracks a click on an outbound link in Google Analytics.
* This function takes a valid URL string as an argument, and uses that URL string
* as the event label.
*/
var trackOutboundLink = function(url) {
   ga('send', 'event', 'outbound', 'click', url, {'hitCallback':
     function () {
     document.location = url;
     }
   });
}


function _gaLt(event){
    var el = event.srcElement || event.target;

    /* Loop up the DOM tree through parent elements if clicked element is not a link (eg: an image inside a link) */
    while(el && (typeof el.tagName == 'undefined' || el.tagName.toLowerCase() != 'a' || !el.href)){
        el = el.parentNode;
    }

    if(el && el.href){
        if(el.href.indexOf(location.host) == -1){ /* external link */
            /* HitCallback function to either open link in either same or new window */
            var hitBack = function(link, target){
                target ? window.open(link, target) : window.location.href = link;
            };
            /* link */
            var link = el.href;
            /* Is target set and not _(self|parent|top)? */
            var target = (el.target && !el.target.match(/^_(self|parent|top)$/i)) ? el.target : false;
            /* send event with callback */
            ga(
                "send", "event", "Outgoing Links", link,
                document.location.pathname + document.location.search,
                {"hitCallback": hitBack(link, target)}
            );

            /* Prevent standard click */
            event.preventDefault ? event.preventDefault() : event.returnValue = !1;
        }

    }
}

/* Attach the event to all clicks in the document after page has loaded */
var w = window;
w.addEventListener ? w.addEventListener("load",function(){document.body.addEventListener("click",_gaLt,!1)},!1)
 : w.attachEvent && w.attachEvent("onload",function(){document.body.attachEvent("onclick",_gaLt)});


    </script>

</head>

<body>


<div class="container" style="padding-bottom: 400px;">
    <div class="page-header">
        <h1>SomaticSeq</h1>

        <h2>An ensemble approach to accurately detect somatic mutations</h2>
    </div>

    <h2>Publication [Open access]</h2>

    <div class="panel panel-default" style="font-family:monospace;">
        <div class="panel-body">
            Manuscript in preparation.
        </div>
    </div>

    <h2>Download SomaticSeq</h2>

    <p>The source code is hosted at <a href="https://github.com/bioinform/somaticseq/" onclick="trackOutboundLink('https://github.com/bioinform/somaticseq'); return false'">https://github.com/bioinform/somaticseq</a>

    <p>The packaged releases can be found at <a href="https://github.com/bioinform/somaticseq/releases"
                                           onclick="trackOutboundLink('https://github.com/bioinform/somaticseq/releases'); return false;">https://github.com/bioinform/somaticseq/releases</a>
    </p>

    <h2>Software Dependencies</h2>
    <ul>
    <li>Python3 and regex package</li>
    <li>R and ada package</li>
    <li>GATK</li>
    <li>snpEFF and snpSIFT</li>
    <li>dbSNP and COSMIC file</li>
    </ul>

    <h2>Running SomaticSeq</h2>
    To use a trained model to predict an existing data set, after running the 5 somatic callers.
    This SomaticSeq workflow takes about 3 hours for an ensmeble call set of 50K calls, and about 6 hours for a call set of half million calls.

    The shell command (VCF file can also be bgzipped, but make sure it has the right .gz extention):
    
    <ul> 
    <li>Somatic SNVs:
    <code>SomaticSeq.SNV.sh -M $PATH/TO/MuTect/variants.vcf -V $PATH/TO/Varscan/variants.snp.vcf -J $PATH/TO/JointSNVMix/variants.vcf -S $PATH/TO/SomaticSniper/variants.vcf -D $PATH/TO/Vardict/variants.vcf -N $PATH/TO/normal.bam -T $PATH/TO/tumor.bam -R $PATH/TO/ada_model_predictor.R -C $PATH/TO/trained.classifier.RData -g human_b37.fasta -c cosmic.b37.v71.vcf -d dbSNP.b37.v141.vcf -s $PATH/TO/snpSift -G $PATH/TO/GenomeAnalysisTK.jar -o $OUTPUT_DIR
    </code>    
    </li>
    <li>Somatic INDELs:
    <code>
    SomaticSeq.INDEL.sh -M $PATH/TO/SomaticIndelDetector/variants.vcf -V $PATH/TO/Varscan/variants.indel.vcf -D $PATH/TO/Vardict/variants.vcf -N $PATH/TO/normal.bam -T $PATH/TO/tumor.bam -R $PATH/TO/ada_model_predictor.R -C $PATH/TO/trained.classifier.RData -g human_b37.fasta -c cosmic.b37.v71.vcf -d dbSNP.b37.v141.vcf -s $PATH/TO/snpSift -G $PATH/TO/GenomeAnalysisTK.jar -o $OUTPUT_DIR
    </code>
    </li>   
    </ul>

    The command-line options for the SomaticSeq scripts are:
    <ul>
    <li><code>-M variants.vcf</code>: VCF file from MuTect</li>
    <li><code>-V variants.vcf</code>: VCF file from VarScan2</li>
    <li><code>-J variants.vcf</code>: JointSNVMix's variant output converted to VCF</li>
    <li><code>-S variants.vcf</code>: VCF file from SomaticSniper</li>
    <li><code>-D [snp|indel].variants.vcf</code>: VarDict's VCF file with only SNVs or INDELs extracted</li>
    <li><code>-N normal.bam</code>: Normal BAM files</li>
    <li><code>-T tumor.bam</code>: Tumor BAM files</li>
    <li><code>-R ada_model_predictor.R</code>: Predictor script in R</li>
    <li><code>-C Trained_Classifier.RData</code>: Trained model/classifier, e.g., from DRAM Challenge</li>
    <li><code>-g human_b37_decoy.fasta</code>: Genome reference FASTA file</li>
    <li><code>-c COSMIC.b37.vcf</code>: COSMIC VCF file</li>
    <li><code>-d dbSNP.v141.b37.vcf</code>: dbSNP VCF file</li>
    <li><code>-s path/to/SnpEFF</code>: snpEFF/snpSift's installation directory containing the executable .jar files</li>
    <li><code>-G path/to/GATK.jar</code>: GATK's java executable file</li>
    <li><code>-o output</code>: Output directory (make sure it exists)</li>
    </ul>

    <h2>References/Tools</h2>

</div>


<!-- Bootstrap core JavaScript
================================================== -->
<!-- Placed at the end of the document so the pages load faster -->
<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"></script>
<script src="http://maxcdn.bootstrapcdn.com/bootstrap/3.2.0/js/bootstrap.min.js"></script>
</body>
