# SEQC2 branch

* This is a special release of the [SEQC2](https://www.fda.gov/science-research/bioinformatics-tools/microarraysequencing-quality-control-maqcseqc#MAQC_IV)'s Somatic Mutation Working Group to establish the cancer reference samples, data and call set. This is based on SomaticSeq v2.8.1, which has been obsolesced by the [current version](https://github.com/bioinform/somaticseq).
* The work to establish the high-confidence somatic mutation call set is published in [Fang, L.T., Zhu, B., Zhao, Y. _et al_. Establishing community reference samples, data and call sets for benchmarking cancer mutation detection using whole-genome sequencing. _Nat Biotechnol_ **39**, 1151-1160 (2021)](https://doi.org/10.1038/s41587-021-00993-6 "Fang LT, et al. Nat Biotechnol (2021)") / [PMID:34504347](http://identifiers.org/pubmed:34504347 "Fang LT, et al. Nat Biotechnol (2021)") / [SharedIt Link](https://rdcu.be/cxs3D "Fang LT, et al. Nat Biotechnol (2021)").

  * The companion paper using the reference data and call set to establish best practice in somatic mutation detection is published in [Xiao, W., Ren, L., Chen, Z. _et al_. Toward best practice in cancer mutation detection with whole-genome and whole-exome sequencing. _Nat Biotechnol_ **39**, 1141-1150 (2021)](https://doi.org/10.1038/s41587-021-00994-5 "Xiao W, et al. Nat Biotechnol (2021)") / [PMID:34504346](http://identifiers.org/pubmed:34504346 "Xiao W, et al. Nat Biotechnol (2021)") / [SharedIt Link](https://rdcu.be/cxASG "Xiao W, et al. Nat Biotechnol (2021)").
  
  * The companion data descriptor paper is published in [Zhao, Y., Fang, L.T., Shen, TW. _et al_. Whole Genome and Exome Sequencing Reference Datasets from A Multi-center and Cross-platform Benchmark Study. _Sci Data_ **8**, 296 (2021)](https://doi.org/10.1038/s41597-021-01077-5 "Zhao Y, et al. Sci Data (2021)").
  
* Many custom scripts were created specifically to complete the SEQC2 project. For regular uses of SomaticSeq, please use the ([main](https://github.com/bioinform/somaticseq)) branch.


### Click [here](docs/seqc2.md) for more SEQC2 details. 


## SEQC2 Somatic Mutation Working Group Data
* Analysis documentation and latest updates about the SEQC2 somatic mutation project can be found at [**https://sites.google.com/view/seqc2**](https://sites.google.com/view/seqc2).
* The [**latest** high-confidence somatic reference call set](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release/latest/) can be found at the NCBI FTP site.
* All the raw sequencing data (FASTQ files) generated by the working group were deposited into [**NCBI SRA database: SRP162370**](https://identifiers.org/ncbi/insdc.sra:SRP162370)
* Some BAM files can be found [here](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/data/WGS/) as well.


----------
## Video Presentations
<table style="text-align: center; width: 100%;">
  <tr>
    <td style="vertical-align: bottom; width: 50%;">Establishing somatic mutation reference call set for the SEQC2 cancer reference samples</td>
    <td style="vertical-align: bottom; width: 50%;">Towards best practice in cancer mutation detection with whole-genome and whole-exome sequencing</td>
  </tr>
  
  <tr>
    <td style="vertical-align: center; width: 50%;"><a href="https://youtu.be/nn0BOAONRe8"><img src="docs/workflow400.png" width=380 height=225 /></a></td>
    <td style="vertical-align: center; width: 50%;"><a href="https://youtu.be/txYQ-UUlvis"><img src="docs/heatmap400.png" width=380 height=223 /></a></td>
  </tr>

</table>

----------


# SEQC2 Custom Scripts

## To re-create reference call set v1.2
* The script we used to generate the latest reference call set is [utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh](utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh).
* The steps are documented with greater details [here](https://sites.google.com/view/seqc2/home/data-analysis/high-confidence-somatic-snv-and-indel-v1-2).
* To re-create v1.2 of the call set with this script, please first download the data files at `https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/analysis/SNVs/cancer_reference_samples_supporting_files/data_to_recreate_call_set_v1-2.tar.gz`, and then unpack it. Read the README.md there. You'll also need to run `download_files.sh` there because you'll need to download additional files. Then you may use our docker image to run the (command specified in the README.md there).

### To re-create the reference call set v1.2, you can run the following commands (requires docker)
```
wget https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/analysis/SNVs/cancer_reference_samples_supporting_files/data_to_recreate_call_set_v1-2.tar.gz
tar -vxf data_to_recreate_call_set_v1-2.tar.gz
cd data_to_recreate_call_set_v1-2
./download_files.sh
docker run --rm -u $UID:$UID --workdir /seqc2 -v $(pwd -P):/seqc2 lethalfang/somaticseq:seqc2_v1.2 /opt/somaticseq/utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh
```


## Analysis scripts
* [seqc2_calculateProbablyOfLowVafCalls.py](utilities/seqc2_calculateProbablyOfLowVafCalls.py) calculates the binomial probability that a somatic mutation would be missed entirely in our high-confidence regions due to low variant allele frequency [Suppl. Sec. 1.7].
  * `utilities/seqc2_calculateProbablyOfLowVafCalls.py -vaf 0.05 0.03 -deep1 175 -deep2 200` shows 
    * At VAF = 0.05, the chances that the variant will be called in at least one out of 21 WGS replicates is approximately 0.99999755 (or it'll be missed entirely at about 2.5 x 10<sup>-6</sup> probability). The chances that it'll receive majority support is about 0.373450. However, the chances that it'll be rescued by higher depth data is approximately 0.994400. 
    * At VAF = 0.03, the three numbers are 0.987777, 0.003745 and 0.845741.
* [estimate_SequencingErrorRate.py](utilities/estimate_SequencingErrorRate.py) estimates overall sequencing error in a BAM file by assuming every base within a BED file is a reference base. The BAM files are normal samples. The BED file is the high-confidence region minus the germline variant positions (Suppl. Table 2).



## Requirement for this SEQC2 release
In order to run `utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh`, the following software need to be installed
* Python 3, plus regex, pysam, numpy, scipy, pandas, xlrd, and openpyxl libraries
* SomaticSeq v3.6.2
* bedtools
* vcftools
* tabix

Alternatively, a docker image is also available at docker hub: `lethalfang/somaticseq:seqc2_v1.2`.
