# SEQC2 branch

* This is a special release for the SEQC-II cancer reference samples project, which was based on SomaticSeq v2.8.1.
* Many custom scripts were created to complete the [SEQC-II](https://www.fda.gov/science-research/bioinformatics-tools/microarraysequencing-quality-control-maqcseqc#MAQC_IV) project.
* Analysis documentation and latest updates about the SEQC-II somatic mutation project can be found at [**https://sites.google.com/view/seqc2**](https://sites.google.com/view/seqc2).
* The [**latest high-confidence reference call set**](http://bit.ly/somatic_call_set) can be found at the NCBI FTP site.
  * The full address for releases: ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release
* Raw sequencing data generated by the working group were deposited into [**NCBI SRA database: SRP162370**](https://trace.ncbi.nlm.nih.gov/Traces/sra/?study=SRP162370)
* For regular uses of SomaticSeq, please use Version 3+ ([master](https://github.com/bioinform/somaticseq)) branch.

# SEQC-II Custom Scripts
* The script we used to generate the latest reference call set is [utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh](utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh).
  * The steps are documented with greater details [here](https://sites.google.com/view/seqc2/home/data-analysis/high-confidence-somatic-snv-and-indel-v1-2).
  * To re-create v1.2 of the call set with this script, please first download the data files needed [here](ftp://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/analysis/cancer_reference_samples_supporting_files/data_to_recreate_call_set_v1-2.tar.gz), and then unpack it. Read the README.md. You'll also need to run `download_files.sh` there because you'll need to download additional files.
* [seqc2_calculateProbablyOfLowVafCalls.py](utilities/seqc2_calculateProbablyOfLowVafCalls.py) calculates the binomial probability that a somatic mutation would be missed entirely in our high-confidence regions due to low variant allele frequency.

# SomaticSeq
* SomaticSeq is an ensemble caller that has the ability to use machine learning to filter out false positives. The detailed documentation is included in the package, located in [docs/Manual.pdf](docs/Manual.pdf "User Manual"). A quick guide can also be found [here](http://bioinform.github.io/somaticseq/).
* SomaticSeq's open-access paper: [Fang LT, Afshar PT, Chhibber A, et al. An ensemble approach to accurately detect somatic mutations using SomaticSeq. Genome Biol. 2015;16:197](http://dx.doi.org/10.1186/s13059-015-0758-2 "Fang LT, Afshar PT, Chhibber A, et al. An ensemble approach to accurately detect somatic mutations using SomaticSeq. Genome Biol. 2015;16:197.").

# Requirement 
In order to run `utilities/makeSeqc2HighConfidenceCallSets/makeTruthSet.sh`, the following software need to be installed
* Python 3, plus regex, pysam, numpy, scipy, pandas, xlrd, and openpyxl libraries
* SomaticSeq v3.6.2
* BEDTools
* VCFTools
* tabix

Alternatively, a docker image is also available at docker hub: `lethalfang/somaticseq:seqc2_v1.2`.

## NOTE
If you have issue navigating FTP site on Chrome, try to enable "Enable support for FTP URLs" in chrome://flags/.
