## Somatic Mutation Working Group of the MAQC/SEQC2 Consortium

When building somatic mutation detection tools and pipelines, it is absolutely necessary to find well characterized references to serve as the "ground truth," against which predicted mutation calls can be evaluated and optimized. The [Genome in a Bottle Consortium](https://www.nist.gov/programs-projects/genome-bottle) has characterized a number of germline samples that have proven extremely valuable. Many germline variant detection pipelines and algorithms have relied on them to train and tune their algorithms, e.g., [DeepVariant](https://github.com/google/deepvariant). However, rigorously characterized genome-wide reference samples and call sets did not exist for somatic mutations until the work by [SEQC2](https://www.fda.gov/science-research/bioinformatics-tools/microarraysequencing-quality-control-maqcseqc#MAQC_IV)'s Somatic Mutation Working Group. The following are some of the working group's papers:

  * [Fang L.T. _et al_. Nat Biotechnol (2021)](https://doi.org/10.1038/s41587-021-00993-6) described the methods to establish the [high-confidence somatic mutation call set and its corresponding high-confidence regions](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/release/latest/) and its corresponding high-confidence regions for a pair of tumor-normal cancer cell lines (HCC1395 vs. HCC1395BL). The genomic DNA was produced in a single batch by ATCC to ensure sample homogeneity. The high-confidence call set was consolidated from 20+ whole genome sequencing replicates from multiple sequencing centers and platforms to a combined 1500X sequencing depth. Three read aligners (i.e., BWA MEM, NovoAlign, and Bowtie2), six somatic mutation callers (i.e., MuTect2, SomaticSniper, VarDict, MuSE, Strelka2, and TNscope), and two machine learning algorithms (SomaticSeq and [NeuSomatic](https://github.com/bioinform/neusomatic)) were used to create the high-confidence call set. The high-confidence call set and regions have been used in a number of companion studies. [DOI:10.1038/s41587-021-00993-6](http://doi.org/10.1038/s41587-021-00993-6) / [PMID:34504347](http://identifiers.org/pubmed/34504347) / [SharedIt Link](https://rdcu.be/cxs3D)

  * [Xiao W. _et al_. Nat Biotechnol (2021)](https://doi.org/10.1038/s41587-021-00994-5) used reference data and call sets described above to investigate how different experimental and bioinformatic factors affect the accuracies of somatic mutation detections in whole-genome and whole-exome sequencings. The experimental factors that were investigated include DNA input amounts, fresh vs. FFPE samples (and different fixation durations), fragmentation methodologies, sequencing libraries, sequencing depths, and tumor purities. Bioinformatic factors included the choices of read aligners, variant callers, quality trimming, error correction, and post-alignment processing (i.e., indel realignment and base quality score recalibration).  [DOI:10.1038/s41587-021-00994-5](http://doi.org/10.1038/s41587-021-00994-5) / [PMID:34504346](http://identifiers.org/pubmed/34504346) / [SharedIt Link](https://rdcu.be/cxASG)

  * [Zhao Y. _et al_. Sci Data (2021)](https://doi.org/10.1038/s41597-021-01077-5) is the data descriptor for all the sequencing data on [SRA:SRP162370](https://identifiers.org/ncbi/insdc.sra:SRP162370) generated by the working group for the pair of tumor-normal reference samples. The multi-center data sets included platforms such as Illumina NovaSeq, Illumina HiSeq, PacBio Sequel, Oxford Nanopore, and 10X Chromium platforms. Sequencing data generated with all the different experimental factors described above are also included. For your convenience, some of the BWA MEM aligned BAM files are directly downloadable on [NCBI's FTP server](https://ftp-trace.ncbi.nlm.nih.gov/ReferenceSamples/seqc/Somatic_Mutation_WG/data/). [DOI:10.1038/s41597-021-01077-5](https://doi.org/10.1038/s41597-021-01077-5) / [PMID:34753956](http://identifiers.org/pubmed/34753956)

  * [Sahraeian S.M.E. _et al_. bioRxiv (2019)](https://doi.org/10.1101/667261) used the sequencing data and high-confidence somatic mutation call set to build deep learning models that accurately detect somatic mutations. 


## Find all of SEQC2's publications:
  * [SEQC2 Collection](https://www.nature.com/collections/seqc2) on Nature Biotechnology
  * [SEQC2 Collection](https://www.biomedcentral.com/collections/SEQC2-article-collection) on Genome Biology
